CXF-8671: Support Jakarta EE 10

karlvr opened the pull request:
Work in progress on supporting Jakarta EE 10.

I have chosen to use the [Jakarta EE BOM](https://mvnrepository.com/artifact/jakarta.platform/jakarta.jakartaee-bom/10.0.0) to manage the versions of all of the Jakarta EE 10 API dependencies in one fell swoop, and to hopefully ensure that CXF is compatible with the specification.

# Remaining

- [x] Jakarta RESTful Web Services 3.1 support in `RuntimeDelegateImpl` and `ResponseImplTest` requires actually implementing the bootstrap APIs
- [x] `EntityPartImpl` and `EntityPartBuilderImpl` need proper integration with registered `MessageBodyReader` and `MessageBodyWriter`
- [X] While a straight `mvn package` succeeds (with the two above classes with stub implementations) the `mvn -Pfastinstall,everything` fails with some missing dependencies (WIP).
- [X] `mvn -Pfastinstall,everything` succeeds
- [ ] Test failures.

karlvr commented:
Current test failures:

```
[INFO] Running org.apache.cxf.jaxrs.utils.ResourceUtilsTest
[INFO] Tests run: 14, Failures: 0, Errors: 0, Skipped: 0, Time elapsed: 0.014 s - in org.apache.cxf.jaxrs.utils.ResourceUtilsTest
[INFO] 
[INFO] Results:
[INFO] 
[ERROR] Failures: 
[ERROR] org.apache.cxf.jaxrs.impl.RetryAfterHeaderProviderTest.test[retry after Sun Nov 03 08:49:37 2002 is duration of PT24884H49M37S from 2000-01-01 00:00:00.000]
[ERROR]   Run 1: RetryAfterHeaderProviderTest.test:73 
Expected: is <PT24884H49M37S>
     but: was <PT24883H49M37S>
[ERROR]   Run 2: RetryAfterHeaderProviderTest.test:73 
Expected: is <PT24884H49M37S>
     but: was <PT24883H49M37S>
[ERROR]   Run 3: RetryAfterHeaderProviderTest.test:73 
Expected: is <PT24884H49M37S>
     but: was <PT24883H49M37S>
[ERROR]   Run 4: RetryAfterHeaderProviderTest.test:73 
Expected: is <PT24884H49M37S>
     but: was <PT24883H49M37S>
[INFO] 
[ERROR] org.apache.cxf.jaxrs.impl.RetryAfterHeaderProviderTest.test[retry after Sun Nov 3 08:49:37 2002 is duration of PT24884H49M37S from 2000-01-01 00:00:00.000]
[ERROR]   Run 1: RetryAfterHeaderProviderTest.test:73 
Expected: is <PT24884H49M37S>
     but: was <PT24883H49M37S>
[ERROR]   Run 2: RetryAfterHeaderProviderTest.test:73 
Expected: is <PT24884H49M37S>
     but: was <PT24883H49M37S>
[ERROR]   Run 3: RetryAfterHeaderProviderTest.test:73 
Expected: is <PT24884H49M37S>
     but: was <PT24883H49M37S>
[ERROR]   Run 4: RetryAfterHeaderProviderTest.test:73 
Expected: is <PT24884H49M37S>
     but: was <PT24883H49M37S>
[INFO] 
[INFO] 
[ERROR] Tests run: 1084, Failures: 2, Errors: 0, Skipped: 1
```

amarkevich commented:
Apache CXF Code Generation Maven2 Plugins .......... FAILURE
https://github.com/apache/cxf/actions/runs/4485519042

karlvr commented:
@amarkevich thank you, that particular problem was resolved in dd20aee697eb61541351a9eb70f08a7984103363. This PR still won't compile due to the `RuntimeDelegateImpl` and `ResponseImplTest` pieces not being done yet. Locally I've implemented with TODOs :-) WIP.

karlvr commented:
The `org.apache.cxf.jaxrs.impl.RetryAfterHeaderProviderTest` test failure I referenced above was because the test did not correctly account for daylight time. I've pushed a fix for that _test_, which would have broken in the main branch.

karlvr commented:
Current test failures that I'm struggling with... I believe all of these failures are related to xml schemas with out a target namespace. It seems that we're not finding the complex type when there's no namespace, but we find it fine when there is. Is anyone able to help understand this?

```
[ERROR] Tests run: 41, Failures: 0, Errors: 12, Skipped: 0, Time elapsed: 3.612 s <<< FAILURE! - in org.apache.cxf.tools.wadlto.jaxrs.JAXRSContainerTest
[ERROR] org.apache.cxf.tools.wadlto.jaxrs.JAXRSContainerTest.testNoTargetNamespace  Time elapsed: 0.283 s  <<< ERROR!
java.lang.RuntimeException: Error compiling schema from WADL : undefined simple or complex type 'testCompositeObject'
	at org.apache.cxf.tools.wadlto.jaxrs.SourceGenerator$InnerErrorListener.error(SourceGenerator.java:2081)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:77)
	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.base/java.lang.reflect.Method.invoke(Method.java:568)
	at org.apache.cxf.common.util.ReflectionInvokationHandler.invoke(ReflectionInvokationHandler.java:88)
	at jdk.proxy2/jdk.proxy2.$Proxy12.error(Unknown Source)
	at com.sun.tools.xjc.api.impl.s2j.SchemaCompilerImpl.error(SchemaCompilerImpl.java:303)
	at com.sun.tools.xjc.util.ErrorReceiverFilter.error(ErrorReceiverFilter.java:55)
	at com.sun.xml.xsom.impl.parser.ParserContext$2.error(ParserContext.java:176)
	at com.sun.xml.xsom.impl.parser.ParserContext$1.reportError(ParserContext.java:154)
	at com.sun.xml.xsom.impl.parser.NGCCRuntimeEx.reportError(NGCCRuntimeEx.java:149)
	at com.sun.xml.xsom.impl.parser.DelayedRef.resolve(DelayedRef.java:80)
...
[ERROR] Errors: 
[ERROR] org.apache.cxf.tools.wadlto.jaxrs.JAXRSContainerTest.testCodeGenDigit
[ERROR]   Run 1: JAXRSContainerTest.testCodeGenDigit:843 » Runtime Error compiling schema from WADL : undefined simple or complex type 'testCompositeObject'
[ERROR]   Run 2: JAXRSContainerTest.testCodeGenDigit:843 » Runtime Error compiling schema from WADL : undefined simple or complex type 'testCompositeObject'
[ERROR]   Run 3: JAXRSContainerTest.testCodeGenDigit:843 » Runtime Error compiling schema from WADL : undefined simple or complex type 'testCompositeObject'
[ERROR]   Run 4: JAXRSContainerTest.testCodeGenDigit:843 » Runtime Error compiling schema from WADL : undefined simple or complex type 'testCompositeObject'
[INFO] 
[ERROR] org.apache.cxf.tools.wadlto.jaxrs.JAXRSContainerTest.testCodeGenHyphen
[ERROR]   Run 1: JAXRSContainerTest.testCodeGenHyphen:821 » Runtime Error compiling schema from WADL : undefined simple or complex type 'testCompositeObject'
[ERROR]   Run 2: JAXRSContainerTest.testCodeGenHyphen:821 » Runtime Error compiling schema from WADL : undefined simple or complex type 'testCompositeObject'
[ERROR]   Run 3: JAXRSContainerTest.testCodeGenHyphen:821 » Runtime Error compiling schema from WADL : undefined simple or complex type 'testCompositeObject'
[ERROR]   Run 4: JAXRSContainerTest.testCodeGenHyphen:821 » Runtime Error compiling schema from WADL : undefined simple or complex type 'testCompositeObject'
[INFO] 
[ERROR] org.apache.cxf.tools.wadlto.jaxrs.JAXRSContainerTest.testNoTargetNamespace
[ERROR]   Run 1: JAXRSContainerTest.testNoTargetNamespace:70 » Runtime Error compiling schema from WADL : undefined simple or complex type 'testCompositeObject'
[ERROR]   Run 2: JAXRSContainerTest.testNoTargetNamespace:70 » Runtime Error compiling schema from WADL : undefined simple or complex type 'testCompositeObject'
[ERROR]   Run 3: JAXRSContainerTest.testNoTargetNamespace:70 » Runtime Error compiling schema from WADL : undefined simple or complex type 'testCompositeObject'
[ERROR]   Run 4: JAXRSContainerTest.testNoTargetNamespace:70 » Runtime Error compiling schema from WADL : undefined simple or complex type 'testCompositeObject'
```

reta commented:
@karlvr apologies I do not have time right now to look at the issue, but I just glance through the changes and the first thing I run into is that dependencies part needs some work, for example `saaj-impl` needs a bump `3.0.0`, probably there are few others too.

[1] https://github.com/apache/cxf/pull/1201/files#diff-b5a06276719e759fe07dfe6f75d781be5f83d2215179d82bdb195ad035348214L179

karlvr commented:
@reta thank you, I've upgraded a few more dependencies ... I still have the namespace issues, and the implementation for `RuntimeDelegateImpl` is still required. I don't think I'm able to proceed from this point as the test failures / implementation required is beyond my knowledge of CXF. When you or someone else gets a chance to take a look it will be great to get past this point and I'll help where I can!

karlvr commented:
Another few wee code contributions... I implemented a basic `SeBootstrap.Configuration` and builder, and I've added stub implementations for the bootstrap APIs in `RuntimeDelegate`... maybe CXF doesn't need to support the bootstrapping immediately, or someone else is definitely better positioned to implement that!

I've taken a stab at the `EntityPart` implementation and builder. The missing piece for me is how to get access to registered `MessageBodyReader` and `Writer` implementations in the context of the `EntityPart.Builder`... I've used `ProviderFactory` with a `null` `Message` for the moment. I am also guessing that somewhere elsewhere in CXF we will need support for `EntityPart`...

reta commented:
> Another few wee code contributions... 

Thanks @karlvr , I think it would have unblocked you a bit on the compilation side, but is it now difficult to track what we need to implement, it would be helpful (to be fair) to get rid of the stubs if possible (we would have to implement the spec to be compliant).

karlvr commented:
@reta yes, I agree. I was holding off on committing them for that reason. It is limited to the `RuntimeDelegateImpl` class. I have listed this issue in the (newly convert to a) checklist in the PR description. Is that enough to keep track of it at this stage? There's also the case of the `EntityPartBuilderImpl` where I've implemented something that I'm pretty sure needs improvements that I'm not sure how to proceed on... I could comment that code out. It's only those two bits that are not just changing versions and search-replace things!

reta commented:
> @reta yes, I agree. I was holding off on committing them for that reason. It is limited to the `RuntimeDelegateImpl` class. I have listed this issue in the (newly convert to a) checklist in the PR description. Is that enough to keep track of it at this stage? 

I was going through the spec and was sketching some very basic implementation (it is very likely some work is repeated). Let's keep it like that but I would be curious to know if you are planning to implement this part of the specification in full so we won't be duplicating the effort. Thank you.

reta commented on bom/pom.xml:
```diff
@@ -20,7 +20,7 @@
     <parent>
         <groupId>org.apache.cxf</groupId>
         <artifactId>cxf</artifactId>
-        <version>4.0.1-SNAPSHOT</version>
```
@karlvr I've bumped the version to 4.1.0 to simplify branch switches, it expanded the changeset quite a bit, apologies for that

karlvr commented:
@reta I've been on a brief holiday! I don't believe I know enough about the ins-and-outs of CXF to contribute to this stage of the PR I think... your implementations look great, and I am really looking forward to seeing it all come together. This is the final piece of my puzzle, or the final so far.

karlvr commented:
@reta Hi, how are you getting on! Is there anything more I can do to help out?

I received a PR on my original branch; it looks good to me but I think you'd be best to consider it as it's more of a behaviour change than the other EE10 conversions I've looked at.

reta commented:
> @reta Hi, how are you getting on! Is there anything more I can do to help out?

Thanks a lot @karlvr , I sadly haven't gotten the support to streamline Jakarta EE 10 development, but it is increasingly difficult to support such a large pull request open for long. I was thinking about creating a separate feature branch so we could work on the migration, merge changes incrementally and have CI running.

[1] https://www.mail-archive.com/dev@cxf.apache.org/msg19805.html

karlvr commented:
@reta I’m totally happy with whatever you need to do to make it easy. I’m happy to rebase on top of HEAD if that’s useful. I fear it’s pretty small. I’m unclear on what is required to continue… I’d be happy to help with a plan and perhaps to execute on it if that seems useful and fun.

reta commented:
> @reta I’m totally happy with whatever you need to do to make it easy. I’m happy to rebase on top of HEAD if that’s useful. I fear it’s pretty small. I’m unclear on what is required to continue… I’d be happy to help with a plan and perhaps to execute on it if that seems useful and fun.

Thanks a lot @karlvr ! So what I've done:
 - created a branch called `CXF-8671` where I just bumped the version to 4.1.0-SNAPSHOT (so all unrelated changes are gone from this pull request)
 - retargeted pull request to `CXF-8671` instead of `main` so we could actually merge it and continue to work iteratively

There are few leftovers on my side (related to JAX-RS) but there is only one thing I wanted to change: the usage of Jakarta EE BOM simplifies a few things but really complicates dependency updates, especially with respect to security fixes. I think we should be better of continuing to manage individual component versions instead (it happens all the time that we need to update components, without need to wait for BOM release which usually takes much longer).

reta commented:
@karlvr I think we are in good shape to merge it into integration branch (and continue work), any objections from your side? thank you!

ffang commented:
Hi @karlvr @reta ,

Thanks for what you have done here, very nice!

And for the Servlet API 6.0 support of EE10, we need to upgrade to use Jetty 12 and Undertow 2.3.x, and I've been working on it. The working on Jetty12 upgrade took me a while as the Jetty12 API changed a lot. I think I've got most issues resolved for Servlet API 6.0 upgrade, and I just pushed my working branch here
https://github.com/apache/cxf/tree/servlet-6

I rebased this branch against branch [CXF-8671](https://issues.apache.org/jira/browse/CXF-8671) you guys have worked on, and I am gonna polish servlet-6 branch a little bit and then send a PR against [CXF-8671](https://issues.apache.org/jira/browse/CXF-8671).

Best Regards
Freeman

reta commented:
> Thanks for what you have done here, very nice!

Thanks @ffang ! I think we got Undertow 2.3.x working already (but not Jetty 12), waiting for @karlvr to give a green light to merge into integration branch so we could move forward!

karlvr commented:
> @karlvr I think we are in good shape to merge it into integration branch (and continue work), any objections from your side? thank you!

@reta absolutely none, I'm excited :-)

rzo1 commented:
Thank you all ;-) - we are also excited to test with a SNAPSHOT ;-)
